File: algo/basic-calculator/calculate.py
class Solution(object):
    def calculate(self, s):
        """
        :type s: str
        :rtype: int
        """
        nums = []
        tmp = 0
        sign = 1
        i = 0
        while i < len(s):
            c = s[i]
            if c >= '0' and c <= '9':
                num = int(c)
                j = i + 1
                while j < len(s) and s[j] >= '0' and s[j] <= '9':
                    num = num *10 + int(s[j])
                    j += 1
                i = j
                tmp += num * sign
            else:
                if c == '+':
                    sign = 1
                elif c == '-':
                    sign = -1
                elif c == '(':
                    nums.append(tmp)
                    nums.append(sign)
                    tmp, sign = 0, 1
                elif c == ')':
                    tmp = tmp * nums.pop() + nums.pop()
                else:
                    pass
                i += 1
        return tmp
                
                
                

File: algo/basic-calculator/calculate.go
func calculate(s string) int {
    nums := []int{}
    tmp, sign := 0, 1
    i := 0
    for i < len(s) {
        c := s[i]
        if c >= '0' && c <= '9' {
            num := int(c - '0')
            j := i + 1
            for j < len(s) && s[j] >= '0' && s[j] <= '9' {
                num = num * 10 + int(s[j] - '0')
                j++
            }
            i = j
            tmp += sign * num
        } else {
            if c == '+' {
                sign = 1
            } else if c == '-' {
                sign = -1
            } else if c == '(' {
                nums = append(nums, tmp, sign)
                tmp, sign = 0, 1
            } else if c == ')' {
                tmp = tmp * nums[len(nums) - 1] + nums[len(nums) - 2]
                nums = nums[:len(nums) - 2]
            } else {
            }
            i++                
        }
    }
    return tmp
}
